File: src/runner/state-manager.test.ts
/**
 * Tests for Brain Runner State Manager
 */

import { describe, it, expect, beforeEach, afterEach } from "bun:test";
import { existsSync, mkdirSync, rmSync, writeFileSync } from "fs";
import { join } from "path";
import { tmpdir } from "os";
import { StateManager } from "./state-manager";
import type { RunnerStats, RunnerStatus, RunningTask } from "./types";

describe("StateManager", () => {
  let testDir: string;
  let manager: StateManager;
  const projectId = "test-project";

  beforeEach(() => {
    // Create a unique temp directory for each test
    testDir = join(tmpdir(), `state-manager-test-${Date.now()}-${Math.random().toString(36).slice(2)}`);
    mkdirSync(testDir, { recursive: true });
    manager = new StateManager(testDir, projectId);
  });

  afterEach(() => {
    // Clean up test directory
    if (existsSync(testDir)) {
      rmSync(testDir, { recursive: true, force: true });
    }
  });

  // Helper to create sample data
  const createSampleTasks = (): RunningTask[] => [
    {
      id: "task-1",
      path: "projects/test/task/abc123.md",
      title: "Test Task 1",
      priority: "high",
      projectId: "test",
      pid: 12345,
      startedAt: "2024-01-15T10:00:00Z",
      isResume: false,
      workdir: "/tmp/work",
    },
    {
      id: "task-2",
      path: "projects/test/task/def456.md",
      title: "Test Task 2",
      priority: "medium",
      projectId: "test",
      pid: 12346,
      paneId: "%5",
      windowName: "test-window",
      startedAt: "2024-01-15T10:05:00Z",
      isResume: true,
      workdir: "/tmp/work2",
    },
  ];

  const createSampleStats = (): RunnerStats => ({
    completed: 5,
    failed: 1,
    totalRuntime: 300000,
  });

  describe("save and load", () => {
    it("saves and loads state round-trip", () => {
      const status: RunnerStatus = "processing";
      const tasks = createSampleTasks();
      const stats = createSampleStats();
      const startedAt = "2024-01-15T09:00:00Z";

      manager.save(status, tasks, stats, startedAt);
      const loaded = manager.load();

      expect(loaded).not.toBeNull();
      expect(loaded!.projectId).toBe(projectId);
      expect(loaded!.status).toBe(status);
      expect(loaded!.startedAt).toBe(startedAt);
      expect(loaded!.runningTasks).toEqual(tasks);
      expect(loaded!.stats).toEqual(stats);
      expect(loaded!.updatedAt).toBeDefined();
    });

    it("load returns null for missing file", () => {
      const result = manager.load();
      expect(result).toBeNull();
    });

    it("load returns null for corrupted JSON", () => {
      const stateFile = join(testDir, `runner-${projectId}.json`);
      writeFileSync(stateFile, "{ invalid json }", "utf-8");

      const result = manager.load();
      expect(result).toBeNull();
    });

    it("load returns null for empty file", () => {
      const stateFile = join(testDir, `runner-${projectId}.json`);
      writeFileSync(stateFile, "", "utf-8");

      const result = manager.load();
      expect(result).toBeNull();
    });
  });

  describe("clear", () => {
    it("removes all state files", () => {
      const status: RunnerStatus = "idle";
      const tasks = createSampleTasks();
      const stats = createSampleStats();

      manager.save(status, tasks, stats, "2024-01-15T09:00:00Z");
      manager.savePid(process.pid);
      manager.saveRunningTasks(tasks);

      const stateFile = join(testDir, `runner-${projectId}.json`);
      const pidFile = join(testDir, `runner-${projectId}.pid`);
      const runningFile = join(testDir, `running-${projectId}.json`);

      expect(existsSync(stateFile)).toBe(true);
      expect(existsSync(pidFile)).toBe(true);
      expect(existsSync(runningFile)).toBe(true);

      manager.clear();

      expect(existsSync(stateFile)).toBe(false);
      expect(existsSync(pidFile)).toBe(false);
      expect(existsSync(runningFile)).toBe(false);
    });

    it("handles missing files gracefully", () => {
      // Should not throw even if files don't exist
      expect(() => manager.clear()).not.toThrow();
    });
  });

  describe("PID file operations", () => {
    it("saves and loads PID", () => {
      const pid = 12345;
      manager.savePid(pid);

      const loaded = manager.loadPid();
      expect(loaded).toBe(pid);
    });

    it("loadPid returns null for missing file", () => {
      const result = manager.loadPid();
      expect(result).toBeNull();
    });

    it("loadPid returns null for invalid content", () => {
      const pidFile = join(testDir, `runner-${projectId}.pid`);
      writeFileSync(pidFile, "not-a-number", "utf-8");

      const result = manager.loadPid();
      expect(result).toBeNull();
    });

    it("clearPid removes the PID file", () => {
      manager.savePid(12345);
      const pidFile = join(testDir, `runner-${projectId}.pid`);
      expect(existsSync(pidFile)).toBe(true);

      manager.clearPid();
      expect(existsSync(pidFile)).toBe(false);
    });

    it("clearPid handles missing file gracefully", () => {
      expect(() => manager.clearPid()).not.toThrow();
    });
  });

  describe("isPidRunning", () => {
    it("returns true for live PID (current process)", () => {
      manager.savePid(process.pid);

      const result = manager.isPidRunning();
      expect(result).toBe(true);
    });

    it("returns false for dead PID", () => {
      // Use a very high PID that's unlikely to exist
      manager.savePid(999999999);

      const result = manager.isPidRunning();
      expect(result).toBe(false);
    });

    it("returns false when no PID file exists", () => {
      const result = manager.isPidRunning();
      expect(result).toBe(false);
    });
  });

  describe("running tasks", () => {
    it("saves and loads running tasks", () => {
      const tasks = createSampleTasks();
      manager.saveRunningTasks(tasks);

      const loaded = manager.loadRunningTasks();
      expect(loaded).toEqual(tasks);
    });

    it("loadRunningTasks returns empty array for missing file", () => {
      const result = manager.loadRunningTasks();
      expect(result).toEqual([]);
    });

    it("loadRunningTasks returns empty array for corrupted JSON", () => {
      const runningFile = join(testDir, `running-${projectId}.json`);
      writeFileSync(runningFile, "{ invalid json }", "utf-8");

      const result = manager.loadRunningTasks();
      expect(result).toEqual([]);
    });
  });

  describe("static utilities", () => {
    describe("findAllRunnerStates", () => {
      it("finds all runner state files", () => {
        // Create multiple runner states
        const manager1 = new StateManager(testDir, "project-a");
        const manager2 = new StateManager(testDir, "project-b");
        const manager3 = new StateManager(testDir, "project-c");

        const stats = createSampleStats();
        manager1.save("idle", [], stats, "2024-01-15T09:00:00Z");
        manager2.save("processing", [], stats, "2024-01-15T10:00:00Z");
        manager3.save("stopped", [], stats, "2024-01-15T11:00:00Z");

        const states = StateManager.findAllRunnerStates(testDir);

        expect(states).toHaveLength(3);
        const projectIds = states.map((s) => s.projectId).sort();
        expect(projectIds).toEqual(["project-a", "project-b", "project-c"]);
      });

      it("returns empty array for non-existent directory", () => {
        const nonExistent = join(testDir, "does-not-exist");
        const states = StateManager.findAllRunnerStates(nonExistent);
        expect(states).toEqual([]);
      });

      it("ignores non-runner files", () => {
        // Create a runner state and some other files
        manager.save("idle", [], createSampleStats(), "2024-01-15T09:00:00Z");
        writeFileSync(join(testDir, "other-file.json"), "{}", "utf-8");
        writeFileSync(join(testDir, "running-test.json"), "[]", "utf-8");
        writeFileSync(join(testDir, "random.txt"), "hello", "utf-8");

        const states = StateManager.findAllRunnerStates(testDir);

        expect(states).toHaveLength(1);
        expect(states[0].projectId).toBe(projectId);
      });
    });

    describe("cleanupStaleStates", () => {
      it("removes states with dead PIDs", () => {
        // Create a state with a dead PID
        const manager1 = new StateManager(testDir, "dead-project");
        manager1.save("processing", [], createSampleStats(), "2024-01-15T09:00:00Z");
        manager1.savePid(999999999); // Dead PID

        // Create a state with live PID
        const manager2 = new StateManager(testDir, "live-project");
        manager2.save("processing", [], createSampleStats(), "2024-01-15T10:00:00Z");
        manager2.savePid(process.pid); // Current process

        const cleaned = StateManager.cleanupStaleStates(testDir);

        expect(cleaned).toBe(1);

        // Dead project should be cleaned
        expect(manager1.load()).toBeNull();

        // Live project should remain
        expect(manager2.load()).not.toBeNull();
      });

      it("returns 0 when no stale states", () => {
        // Create only live states
        const manager1 = new StateManager(testDir, "live-project");
        manager1.save("processing", [], createSampleStats(), "2024-01-15T09:00:00Z");
        manager1.savePid(process.pid);

        const cleaned = StateManager.cleanupStaleStates(testDir);
        expect(cleaned).toBe(0);
      });

      it("handles empty directory", () => {
        const cleaned = StateManager.cleanupStaleStates(testDir);
        expect(cleaned).toBe(0);
      });
    });
  });
});

File: internal/runner/state_manager_test.go
package runner

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"
)

// ---------------------------------------------------------------------------
// StateManager — Save and Load
// ---------------------------------------------------------------------------

func TestStateManager_SaveAndLoad(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "test-project")

	tasks := []RunningTask{
		{
			ID:        "task1",
			Path:      "projects/test/task/abc12345.md",
			Title:     "Test Task",
			Priority:  "high",
			ProjectID: "test-project",
			PID:       12345,
			StartedAt: time.Now(),
		},
	}
	stats := RunnerStats{Completed: 5, Failed: 1, TotalRuntime: 60000}
	startedAt := time.Now()

	sm.Save(RunnerStatusProcessing, tasks, stats, startedAt)

	loaded := sm.Load()
	if loaded == nil {
		t.Fatal("Load returned nil, expected state")
	}

	if loaded.ProjectID != "test-project" {
		t.Errorf("ProjectID = %q, want %q", loaded.ProjectID, "test-project")
	}
	if loaded.Status != RunnerStatusProcessing {
		t.Errorf("Status = %q, want %q", loaded.Status, RunnerStatusProcessing)
	}
	if len(loaded.RunningTasks) != 1 {
		t.Fatalf("RunningTasks len = %d, want 1", len(loaded.RunningTasks))
	}
	if loaded.RunningTasks[0].ID != "task1" {
		t.Errorf("RunningTasks[0].ID = %q, want %q", loaded.RunningTasks[0].ID, "task1")
	}
	if loaded.Stats.Completed != 5 {
		t.Errorf("Stats.Completed = %d, want 5", loaded.Stats.Completed)
	}
	if loaded.Stats.Failed != 1 {
		t.Errorf("Stats.Failed = %d, want 1", loaded.Stats.Failed)
	}
}

func TestStateManager_Load_NoFile(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "nonexistent")

	loaded := sm.Load()
	if loaded != nil {
		t.Errorf("Load returned %v, want nil for missing file", loaded)
	}
}

func TestStateManager_Load_CorruptedJSON(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "corrupt")

	// Write invalid JSON to the state file
	stateFile := filepath.Join(dir, "runner-corrupt.json")
	if err := os.WriteFile(stateFile, []byte("{invalid json"), 0o644); err != nil {
		t.Fatalf("write corrupt file: %v", err)
	}

	loaded := sm.Load()
	if loaded != nil {
		t.Errorf("Load returned %v, want nil for corrupted JSON", loaded)
	}
}

// ---------------------------------------------------------------------------
// StateManager — Clear
// ---------------------------------------------------------------------------

func TestStateManager_Clear(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "clearme")

	// Create all three files
	sm.Save(RunnerStatusIdle, nil, RunnerStats{}, time.Now())
	sm.SavePid(os.Getpid())
	sm.SaveRunningTasks([]RunningTask{{ID: "t1"}})

	// Verify files exist
	for _, name := range []string{"runner-clearme.json", "runner-clearme.pid", "running-clearme.json"} {
		if _, err := os.Stat(filepath.Join(dir, name)); err != nil {
			t.Fatalf("expected file %s to exist before clear", name)
		}
	}

	sm.Clear()

	// Verify files are gone
	for _, name := range []string{"runner-clearme.json", "runner-clearme.pid", "running-clearme.json"} {
		if _, err := os.Stat(filepath.Join(dir, name)); !os.IsNotExist(err) {
			t.Errorf("expected file %s to be removed after clear", name)
		}
	}
}

func TestStateManager_Clear_NoFiles(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "empty")

	// Should not panic when files don't exist
	sm.Clear()
}

// ---------------------------------------------------------------------------
// StateManager — PID Management
// ---------------------------------------------------------------------------

func TestStateManager_SaveAndLoadPid(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "pidtest")

	sm.SavePid(42)

	pid := sm.LoadPid()
	if pid == nil {
		t.Fatal("LoadPid returned nil, expected 42")
	}
	if *pid != 42 {
		t.Errorf("LoadPid = %d, want 42", *pid)
	}
}

func TestStateManager_LoadPid_NoFile(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "nopid")

	pid := sm.LoadPid()
	if pid != nil {
		t.Errorf("LoadPid returned %v, want nil for missing file", pid)
	}
}

func TestStateManager_LoadPid_InvalidContent(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "badpid")

	pidFile := filepath.Join(dir, "runner-badpid.pid")
	if err := os.WriteFile(pidFile, []byte("not-a-number"), 0o644); err != nil {
		t.Fatalf("write pid file: %v", err)
	}

	pid := sm.LoadPid()
	if pid != nil {
		t.Errorf("LoadPid returned %v, want nil for invalid content", pid)
	}
}

func TestStateManager_ClearPid(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "clearpid")

	sm.SavePid(99)
	sm.ClearPid()

	pid := sm.LoadPid()
	if pid != nil {
		t.Errorf("LoadPid returned %v after ClearPid, want nil", pid)
	}
}

// ---------------------------------------------------------------------------
// StateManager — IsPidRunning
// ---------------------------------------------------------------------------

func TestStateManager_IsPidRunning_CurrentProcess(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "running")

	// Save our own PID — it should be running
	sm.SavePid(os.Getpid())

	if !sm.IsPidRunning() {
		t.Error("IsPidRunning = false for current process, want true")
	}
}

func TestStateManager_IsPidRunning_DeadProcess(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "dead")

	// PID 99999999 almost certainly doesn't exist
	sm.SavePid(99999999)

	if sm.IsPidRunning() {
		t.Error("IsPidRunning = true for dead PID, want false")
	}
}

func TestStateManager_IsPidRunning_NoPidFile(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "nopidfile")

	if sm.IsPidRunning() {
		t.Error("IsPidRunning = true with no PID file, want false")
	}
}

// ---------------------------------------------------------------------------
// StateManager — Running Tasks
// ---------------------------------------------------------------------------

func TestStateManager_SaveAndLoadRunningTasks(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "tasks")

	tasks := []RunningTask{
		{ID: "t1", Title: "Task 1", ProjectID: "proj"},
		{ID: "t2", Title: "Task 2", ProjectID: "proj"},
	}

	sm.SaveRunningTasks(tasks)

	loaded := sm.LoadRunningTasks()
	if len(loaded) != 2 {
		t.Fatalf("LoadRunningTasks len = %d, want 2", len(loaded))
	}
	if loaded[0].ID != "t1" {
		t.Errorf("loaded[0].ID = %q, want %q", loaded[0].ID, "t1")
	}
	if loaded[1].ID != "t2" {
		t.Errorf("loaded[1].ID = %q, want %q", loaded[1].ID, "t2")
	}
}

func TestStateManager_LoadRunningTasks_NoFile(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "notasks")

	loaded := sm.LoadRunningTasks()
	if len(loaded) != 0 {
		t.Errorf("LoadRunningTasks len = %d, want 0 for missing file", len(loaded))
	}
}

func TestStateManager_LoadRunningTasks_CorruptedJSON(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "corrupttasks")

	tasksFile := filepath.Join(dir, "running-corrupttasks.json")
	if err := os.WriteFile(tasksFile, []byte("not json"), 0o644); err != nil {
		t.Fatalf("write tasks file: %v", err)
	}

	loaded := sm.LoadRunningTasks()
	if len(loaded) != 0 {
		t.Errorf("LoadRunningTasks len = %d, want 0 for corrupted JSON", len(loaded))
	}
}

// ---------------------------------------------------------------------------
// StateManager — FindAllRunnerStates (static)
// ---------------------------------------------------------------------------

func TestFindAllRunnerStates(t *testing.T) {
	dir := t.TempDir()

	// Create some runner state files
	for _, name := range []string{"runner-proj1.json", "runner-proj2.json", "other-file.txt", "running-proj1.json"} {
		if err := os.WriteFile(filepath.Join(dir, name), []byte("{}"), 0o644); err != nil {
			t.Fatalf("write file %s: %v", name, err)
		}
	}

	states := FindAllRunnerStates(dir)
	if len(states) != 2 {
		t.Fatalf("FindAllRunnerStates len = %d, want 2", len(states))
	}

	// Check project IDs (order may vary)
	ids := map[string]bool{}
	for _, s := range states {
		ids[s.ProjectID] = true
	}
	if !ids["proj1"] || !ids["proj2"] {
		t.Errorf("expected proj1 and proj2, got %v", ids)
	}
}

func TestFindAllRunnerStates_EmptyDir(t *testing.T) {
	dir := t.TempDir()

	states := FindAllRunnerStates(dir)
	if len(states) != 0 {
		t.Errorf("FindAllRunnerStates len = %d, want 0 for empty dir", len(states))
	}
}

func TestFindAllRunnerStates_NonexistentDir(t *testing.T) {
	states := FindAllRunnerStates("/nonexistent/path/that/does/not/exist")
	if len(states) != 0 {
		t.Errorf("FindAllRunnerStates len = %d, want 0 for nonexistent dir", len(states))
	}
}

// ---------------------------------------------------------------------------
// StateManager — CleanupStaleStates (static)
// ---------------------------------------------------------------------------

func TestCleanupStaleStates(t *testing.T) {
	dir := t.TempDir()

	// Create a state file with a dead PID
	sm := NewStateManager(dir, "stale")
	sm.Save(RunnerStatusIdle, nil, RunnerStats{}, time.Now())
	sm.SavePid(99999999) // dead PID

	// Create a state file with our own PID (alive)
	sm2 := NewStateManager(dir, "alive")
	sm2.Save(RunnerStatusProcessing, nil, RunnerStats{}, time.Now())
	sm2.SavePid(os.Getpid())

	cleaned := CleanupStaleStates(dir)
	if cleaned != 1 {
		t.Errorf("CleanupStaleStates = %d, want 1", cleaned)
	}

	// Stale state should be gone
	if sm.Load() != nil {
		t.Error("stale state should have been cleaned up")
	}

	// Alive state should remain
	if sm2.Load() == nil {
		t.Error("alive state should not have been cleaned up")
	}
}

// ---------------------------------------------------------------------------
// StateManager — JSON serialization format
// ---------------------------------------------------------------------------

func TestStateManager_SaveFormat(t *testing.T) {
	dir := t.TempDir()
	sm := NewStateManager(dir, "format")

	startedAt := time.Date(2025, 1, 15, 10, 30, 0, 0, time.UTC)
	sm.Save(RunnerStatusPolling, nil, RunnerStats{Completed: 3}, startedAt)

	// Read raw file and verify it's valid JSON
	data, err := os.ReadFile(filepath.Join(dir, "runner-format.json"))
	if err != nil {
		t.Fatalf("read state file: %v", err)
	}

	var raw map[string]interface{}
	if err := json.Unmarshal(data, &raw); err != nil {
		t.Fatalf("state file is not valid JSON: %v", err)
	}

	if raw["projectId"] != "format" {
		t.Errorf("projectId = %v, want %q", raw["projectId"], "format")
	}
	if raw["status"] != "polling" {
		t.Errorf("status = %v, want %q", raw["status"], "polling")
	}
}
